File: p001/p001.scala
// If we list all the natural numbers below 10 that are multiples of 3 or 5,
// we get 3, 5, 6 and 9. The sum of these multiples is 23.

// Find the sum of all the multiples of 3 or 5 below 1000.

import scala.language.postfixOps

def multiple3or5(n: Int): Boolean = {
  n % 3 == 0 || n % 5 == 0
}

def sumAllMultiplesOf3Or5BelowN(n: Int): Int = {
  (1 to n-1) filter multiple3or5 sum
}

sumAllMultiplesOf3Or5BelowN(10)    // 23

sumAllMultiplesOf3Or5BelowN(1000)  // 233168

File: p001/p001.rs
// If we list all the natural numbers below 10 that are multiples of 3 or 5,
// we get 3, 5, 6 and 9. The sum of these multiples is 23.

// Find the sum of all the multiples of 3 or 5 below 1000.

fn sum_all_multiples_of_3_or_5_below_n(n: i32) -> i32 {
  (1..n).filter(|x| x % 3 == 0 || x % 5 == 0).fold(0, |sum, x| sum + x)
}

fn main() {
  println!("{}", sum_all_multiples_of_3_or_5_below_n(10));    // 23
  println!("{}", sum_all_multiples_of_3_or_5_below_n(1000));  // 233168
}
